plutus-benchmark/bls-benchmarks/src/PlutusBenchmark/BLS12_381/Common.hs

```diff
@@ -0,0 +1,349 @@
+checkUncompressAndAddG1_Haskell n =
+    let l = listOfSizedByteStrings 100 n
+        points = map (Tx.bls12_381_G1_compress . Tx.bls12_381_G1_hashToGroup . toBuiltin) l
+        result1 = uncompressAndAddG1 points
+        result2 = uncompressAndAddG1 (reverse points)
+    in do
+      printf "%s\n" (show result1)
+      printf "%s\n" (show result2)
+
+-- Deserialise a list of compressed G1 points and add them all together
+{-# INLINABLE uncompressAndAddG2 #-}
+uncompressAndAddG2 :: [BuiltinByteString] -> BuiltinBLS12_381_G2_Element
+uncompressAndAddG2 [] = error ()
+uncompressAndAddG2 (p:ps) =
+    go ps (Tx.bls12_381_G2_uncompress p)
+    where go [] acc     = acc
+          go (q:qs) acc = go qs $ Tx.bls12_381_G2_add (Tx.bls12_381_G2_uncompress q) acc
+
+mkUncompressAndAddG2Script :: [ByteString] -> UProg
+mkUncompressAndAddG2Script l =
+    let points = map (Tx.bls12_381_G2_compress . Tx.bls12_381_G2_hashToGroup . toBuiltin) l
+    in Tx.getPlcNoAnn $ $$(Tx.compile [|| uncompressAndAddG2 ||]) `Tx.unsafeApplyCode` Tx.liftCode points
+
+-- Check that point addition is commutative in G2
+checkUncompressAndAddG2_Haskell :: Integer -> IO ()
+checkUncompressAndAddG2_Haskell n =
+    let l = listOfSizedByteStrings 100 n
+        points = map (Tx.bls12_381_G2_compress . Tx.bls12_381_G2_hashToGroup . toBuiltin) l
+        result1 = uncompressAndAddG2 points
+        result2 = uncompressAndAddG2 (reverse points)
+    in do
+      printf "%s\n" (show result1)
+      printf "%s\n" (show result2)
+
+-- Pairing operations
+
+-- Take two points p1 and p2 in G1 and two points q1 and q2 in G2, apply the
+-- Miller loop to (p1,q1) and (p2,q2), and then call finalVerify on the results.
+{-# INLINABLE runPairingFunctions #-}
+runPairingFunctions
+    :: Tx.BuiltinBLS12_381_G1_Element
+    -> Tx.BuiltinBLS12_381_G2_Element
+    -> Tx.BuiltinBLS12_381_G1_Element
+    -> Tx.BuiltinBLS12_381_G2_Element
+    -> Bool
+runPairingFunctions p1 q1 p2 q2 =
+    let r1 = Tx.bls12_381_millerLoop p1 q1
+        r2 = Tx.bls12_381_millerLoop p2 q2
+    in Tx.bls12_381_finalVerify r1 r2
+
+mkPairingScript
+    :: BuiltinBLS12_381_G1_Element
+    -> BuiltinBLS12_381_G2_Element
+    -> BuiltinBLS12_381_G1_Element
+    -> BuiltinBLS12_381_G2_Element
+    -> UProg
+mkPairingScript p1 q1 p2 q2 =
+    Tx.getPlcNoAnn $ $$(Tx.compile [|| runPairingFunctions ||])
+          `Tx.unsafeApplyCode` Tx.liftCode p1
+          `Tx.unsafeApplyCode` Tx.liftCode q1
+          `Tx.unsafeApplyCode` Tx.liftCode p2
+          `Tx.unsafeApplyCode` Tx.liftCode q2
+
+---------------- Groth16 verification ----------------
+
+{- | An example of the on-chain computation required for verification of a Groth16
+ proof.  The data here is derived from
+ https://github.com/achimcc/groth16-example/blob/main/src/lib.rs -}
+
+-- Wrappers for serialised group elements for slightly better type safety
+newtype SerialisedG1Element = SerialisedG1Element { g1 :: BuiltinByteString }
```

Comment:
Wait, are these serialized or compressed? Where do we make them not serialized?

Reply:
They're compressed.  I think I changed the name at come point and forgot to change it here.  I've fixed that now.  Well spotted.